File: src/design_patterns/structural/proxy/javascript/proxy.js
class Subject {
    request() {}
  }
  
  class RealSubject extends Subject {
    request() {
      console.log("RealSubject: Handling request.");
    }
  }
  
  class Proxy extends Subject {
    constructor(realSubject) {
      super();
      this._realSubject = realSubject;
    }
  
    request() {
      if (this.checkAccess()) {
        this._realSubject.request();
        this.logAccess();
      }
    }
  
    checkAccess() {
      console.log("Proxy: Checking access before handling request.");
      return true;
    }
  
    logAccess() {
      console.log("Proxy: Logging the time of request.");
    }
  }
  
  // Usage
  const realSubject = new RealSubject();
  const proxy = new Proxy(realSubject);
  
  proxy.request();
  

File: src/design_patterns/structural/proxy/c#/proxy.cs
using System;

interface ISubject
{
    void Request();
}

class RealSubject : ISubject
{
    public void Request()
    {
        Console.WriteLine("RealSubject: Handling request.");
    }
}

class Proxy : ISubject
{
    private readonly RealSubject _realSubject;

    public Proxy(RealSubject realSubject)
    {
        _realSubject = realSubject;
    }

    public void Request()
    {
        if (CheckAccess())
        {
            _realSubject.Request();
            LogAccess();
        }
    }

    private bool CheckAccess()
    {
        Console.WriteLine("Proxy: Checking access before handling request.");
        return true;
    }

    private void LogAccess()
    {
        Console.WriteLine("Proxy: Logging the time of request.");
    }
}

/*
// Usage
RealSubject realSubject = new RealSubject();
Proxy proxy = new Proxy(realSubject);

proxy.Request();
*/